### ISSUE TITLE
Chapter 5. Replication

### ISSUE THREAD

**Issue by Zhenyu-J** [Enele2/DDIA-notes#2]
## Replication purposes
   - High availability:
        - Keeping the system running, even when one machine (or several machines, or an entire datacenter) goes down
    - Disconnected operation:
        - Allowing an application to continue working when there is a network interruption
    - Latency:
        - Placing data geographically close to users, so that users can interact with it faster
    - Scalability:
        - Being able to handle a higher volume of reads than a single machine could handle, by performing reads on replicas
- Single-leader/multi-leader/leaderless
    - Single-leader replication(async): easy to understand and there is no conflict resolution.
        - Clients send all writes to a single node (the leader), which sends a stream of data change events to the other replicas (followers). Reads can be performed on any replica, but reads from followers might be stale.
    - Multi-leader replication(MySQL, PostgreSQL by default, Etherpad, Google Doc, offline editing): more robust in the presence of faulty nodes, network interruptions, and latency spikes
        - Clients send each write to several leader nodes, any of which can accept writes. The leaders send streams of data change events to each other and to any follower nodes.
    - Leaderless replication:
        - Clients send each write to several nodes, and read from several nodes in parallel in order to detect and correct nodes with stale data.
## Single-leader sync/async
   - Synchronous: consistence, but slow, a single point of failure will block whole system
        - Usually use semi-synchronous, the leader has one synchronous follower.
    - Asynchronous:
        - can be fast, write will not be blocked even followers fall behind
        - However, if a leader fails and you promote an asynchronous updated follower to be the new leader, recently committed data may be lost.
        - Eventual consistency
## Consistency models -> to prevent replication lag
   - Read-after-write consistency
        - Users should always see data that they submitted themselves.
        - How: When reading something that the user may have modified, read it from the leader; otherwise, read it from a follower.
        - Example: read owner’s self profile
    - Monotonic reads
    - Consistent prefix reads
## Multi-leader replication
   - Use cases for multi-leader replication:
        - Multi-datacenter operation
            - Every write can be processed in the local datacenter and is replicated asynchronously to the other datacenters. Thus, the inter-datacenter network delay is hidden from users, which means the perceived performance may be better.
        - Clients with offline operation:
            - Works for applications that need to continue to work while it is disconnected from the internet
            - Every device has a local database that acts as a leader, and there is an asynchronous multi-leader replication process
        - Collaborative editing
            - For faster collaboration, make the unit of change very small & avoid locking
    - Handling write conflicts
        - Synchronous vs asynchronous conflict detection
            - If you want synchronous conflict detection, you might as well just use single-leader replication.
            - In a multi-leader setup, both writes are successful, and the conflict is only detected asynchronously.
        - Conflict avoidance (frequently recommended)
            - If the application can ensure that all writes for a particular record go through the same leader, then conflicts cannot occur.
            - Eg, requests from a particular user are always routed to the same datacenter and use the leader in that datacenter for reading and writing.
            - However, sometimes you might want to change the designated leader for a record because one datacenter has failed. You have to deal with the possibility of concurrent writes on different leaders.
        - Converging toward a consistent state
            - Every replication scheme must ensure that the data is eventually the same in all replicas.
            - Eg, give each write a unique ID, pick the write with the highest ID as the winner, and throw away the other writes. If a timestamp is used, it’s know as “last write win (LWW)”
            - Eg, record the conflict and preserve all information, and resolve the conflict at some later time (perhaps by prompting the user).
        - Custom conflict resolution logic
            - On write
            - On read
    - Multileader replication topologies
        - Circular topology (MySQL by default)
        - Star topology
        - All-to-all topology (most general)
            - Benefit: avoid single point of failure
            - Might have consistent prefix read problem - version vectors to fix it
            - Infinite replication loop prevention. 同上
## Leaderless replication
- ![5-11](https://github.com/Zhenyu-J/DDIA-notes/raw/main/img/5-11.png)
    
    - Leaderless (Dynamo):
        - Benefit: unlike a leader database, that coordinator does not enforce a particular ordering of writes. Can keep writing to the DB when a node is down
        - The benefit of leaderless is that as long as sufficient nodes complete a write, you can successfully continue. When nodes come back up, some of the data is stale.
        - Clients detect stale reads by reading from multiple nodes too
        - Catching up on stale data from missed writes
            - Read repair - when a client detects a stale read, it writes the newer value back
            - Anti-entropy process - background process that searches for differences between replicas and corrects them, No guarantee of order or time until repair.
    - Quorums for reading and writing
        - w + r > n, commonly set n = 3
        - A common choice is to make n an odd number and to set w = r = (n + 1) / 2 (round up).
        - A workload with few writes and many reads, setting w = n and r = 1, this makes reads faster, but just one failed node causes all DB writes to fail.
    - Detecting concurrent writes
        - Last write wins (discarding concurrent writes)
        - The “happens-before” relationship and concurrency
        - Merging concurrently written values
        - Version vectors

### Reference
Zookeeper vs load balancer

In short, zookeeper acts like a cheaper load balancer for registry (registry: 管理master/slave如何对接, zookeeper 相当于小组件)

[DDIA: chapter 5](https://docs.google.com/document/d/1T8Y4DNrJXr9Cxemz2J9SqxCMA1M115G7o8_-ggMUz9U/edit#heading=h.w908ls2ppawf)

https://zookeeper.apache.org/doc/current/index.html

https://medium.com/@nrkivar/zookeeper-de7259a45681#:~:text=Zookeeper%20is%20registry%20for%20large,of%20workload%20across%20multiple%20servers.

https://medium.com/@sureshpodeti/distributed-coordination-service-zookeeper-309ab626097d

https://ssudan16.medium.com/apache-zookeeper-internals-7b063f9d74ac
